### PR TITLE
add xcorr_detector function

### PR DESCRIPTION
**Pull request by trichter**
I added a new `xcorr_detector` function which is based on `correlate_template` and more flexible than `templates_max_similarity`.

* return list of all detections
* Streams dont have to be aligned
* correct normalization by using correlate_template

### PR Checklist
- [x] Correct base branch selected? `master` for new features, `maintenance_...` for bug fixes
- [x] All tests still pass.
- [x] Any new features or fixed regressions are be covered via new tests.
- [x] Any new or changed features have are fully documented.
- [x] Significant changes have been added to `CHANGELOG.txt` .
- [x] Add tutorial
- [x] Add figure to gallery

### CODE REVIEW

**Comment by trichter** [comment]
+DOCS

File: obspy/signal/cross_correlation.py [hunk 1]
```diff
@@ -849,6 +850,123 @@ def templates_max_similarity(st, time, streams_templates):
...
+    :param kwargs: kwargs are passed to correlate_template function
+
+    :return: list of UTCDateTime objects with detections and
+        dictionary with correlations {seed_id: numpy.array}
+    """
+    # preparations
+    stream.merge(fill_value=0)
+    if len({tr.stats.sampling_rate for tr in stream + template}) > 1:
+        raise ValueError('Traces have different sampling rate')
+    ids = {tr.id for tr in stream} | {tr.id for tr in template}
```
Comment (d-chambers) [inline obspy/signal/cross_correlation.py]: I think you mean `ids = {tr.id for tr in stream} & {tr.id for tr in template}`

File: obspy/signal/cross_correlation.py [hunk 2]
```diff
@@ -849,6 +850,123 @@ def templates_max_similarity(st, time, streams_templates):
...
+    if len(ids) == 0:
+        raise ValueError('No traces with matching ids in template and stream')
+    stream.traces = [tr for tr in stream if tr.id in ids]
+    template.traces = [tr for tr in template if tr.id in ids]
```
Comment (d-chambers) [inline obspy/signal/cross_correlation.py]: These change the `stream` instances in place, perhaps worth copying first?

**Comment by megies** [comment]
@trichter would be good to have a docs Tutorial page with an example use case. If you need data for it we can put it in http://examples.obspy.org/ via https://github.com/obspy/examples

File: obspy/signal/cross_correlation.py (hunk 2 above)
Comment (megies) [reply obspy/signal/cross_correlation.py]: A shallow copy might even suffice (without havin looked through all code, depends on whether that stream is just throwaway inside a block or returned actually)

**Comment by megies** [comment]
Rebased on current master and force-pushed so that we have fresh CI results tomorrow.

File: obspy/signal/cross_correlation.py (hunk 1 above)
Comment (trichter) [reply obspy/signal/cross_correlation.py]: Yea, you are right

**Comment by trichter** [comment]
> This looks cool @trichter. I like the idea of having a basic xcorr detector in obspy. Of course, if users outgrow it they can switch over to an external package like EQCorrScan, so it doesn't need to be wildly efficient or comprehensive.

Thanks, that was exactly my idea. Also it is just natural now that the correlate_template function is available.

>     1. The `xcorr_detector` function  is very long and doing a lot of things. Maybe it could be broken up into smaller functions? I am thinking of `_prep_streams` function to do all the trimming/seed_id filtering, `_correlate_streams` to preform the actual correlations, then a distinct plotting function (similar to `plot_trigger`). It might also be better to have a function that returns the ccs dictionary and a different function that picks out the detections from it, in the same spirit as the other detectors in obspy.  If the output could somehow be used with `coincidence_trigger` that would be wonderful.

The preparation is not that much. It is also a lot about setting default values for optional argument. If this is moved to a separate function the docs would be split up, too.
But I would be OK with spliting up the function into two: correlate_stream_template and xcorr_detector. The first function could also return a stream with correlations, which could be used with the coincidence_trigger, too. I don't know how useful this is. Still, I like the approach of a single function a bit more. Other opinions?

>     2. Filling all the gaps with 0s on the input stream may not be what the user should do in all cases. Might be better to just require there are no gaps in the input stream to force the user to address data quality issues rather than returning potentially surprising results.

That sounds quiet reasonable.

>     3. This great addition could really use a tutorial page to add visibility. Maybe it belongs as a section in the Tirgger/Picker Tutorial, or perhaps it would merit a page of its own.

Good idea, I will create a short tutorial, just wanted to get some opinions first.

**Comment by megies** [comment]
>>  1. The `xcorr_detector` function ....

> The preparation is not that much. It is also a lot about setting default values for optional argument. If this is moved to a separate function the docs would be split up, too.
> But I would be OK with spliting up the function into two: correlate_stream_template and xcorr_detector. The first function could also return a stream with correlations, which could be used with the coincidence_trigger, too. I don't know how useful this is. Still, I like the approach of a single function a bit more. Other opinions?

Time is tight, kind of. So I'd be OK with leaving it inside a god function for now. Refactoring can still be done afterwards if anybody feels a need to reuse part of the funtionality.


>> 2. Filling all the gaps with 0s ..

> That sounds quiet reasonable.

I agree, usually I'd prefer raising an exception on strange/unexpected input. Users should be forced to take care of their input rather than loads of obscure automagic.



>> 3. This great addition could really use a tutorial page

> Good idea, I will create a short tutorial, just wanted to get some opinions first.

:rocket:

**Comment by megies** [comment]
Circle CI is failing, yay! That means rebasing (all the open PRs) was already useful for something (to see this issue over all the test failure noise). :rocket:

**Comment by trichter** [comment]
Can somebody have a look? Later today I will create the tutorial.

**Comment by trichter** [comment]
Links to the docs:
[correlate_stream_template](http://docs.obspy.org/pull-requests/2315/packages/autogen/obspy.signal.cross_correlation.correlate_stream_template.html#obspy.signal.cross_correlation.correlate_stream_template)
[http://docs.obspy.org/pull-requests/2315/packages/autogen/obspy.signal.cross_correlation.similarity_detector.html#obspy.signal.cross_correlation.similarity_detector](similarity_detector)
[http://docs.obspy.org/pull-requests/2315/tutorial/code_snippets/xcorr_detector.html](Tutorial)

**Comment by ThomasLecocq** [comment]
@trichter this http://docs.obspy.org/pull-requests/2315/tutorial/code_snippets/xcorr_detector.html doesn't work ?

**Comment by trichter** [comment]
It's working now, the doc buildbot had not yet finished. I still have to figure out how to make the plots appear. Consider the tutorial as WIP.

**Comment by trichter** [comment]
I rebased and force-pushed

File: misc/docs/source/tutorial/code_snippets/xcorr_detector.rst [hunk 3]
```diff
@@ -7,18 +7,25 @@ This code snippets shows how to use the functions
...
 
     from obspy import read, UTCDateTime as UTC
     from obspy.signal.cross_correlation import correlate_stream_template, similarity_detector
 
     template = read('https://examples.obspy.org/IC.MDJ.2013.043.mseed')
     template.filter('bandpass', freqmin=0.5, freqmax=2)
     template.plot()
+
+.. plot::
+    :context:
```
Comment (megies) [inline misc/docs/source/tutorial/code_snippets/xcorr_detector.rst]: from the Sphinx docs it's quite unclear how long this context stays active..

Comment (trichter) [reply misc/docs/source/tutorial/code_snippets/xcorr_detector.rst]: I get your point. If I use :context: reset in the first code cell that should be OK.

**Comment by trichter** [comment]
@d-chambers I think this PR is ready for review. Could you have a look again?

File: obspy/signal/cross_correlation.py [hunk 4]
```diff
@@ -849,6 +850,196 @@ def templates_max_similarity(st, time, streams_templates):
...
+        channels.
+    :param mean_thresshold: Similarity value to trigger a detection.
+    :param holdon: Hold time in seconds, only the detection with largest
+        similarity value in that time span is returned.
+    :param holdoff: Hold time in seconds, detections are
+        separated by this time span.
+    :param similarity: The similarity Trace can be directly specified.
+        Otherwise the similarity is calculated as the mean of
+        cross-correlations.
+    :param plot_detections: Plot detections together with the data of the
```
[the remark this answers is not in the document]
Comment (trichter) [reply obspy/signal/cross_correlation.py]: Good point.

File: obspy/signal/cross_correlation.py [hunk 5]
```diff
@@ -849,6 +850,196 @@ def templates_max_similarity(st, time, streams_templates):
...
+    dt = similarity.stats.delta
+    holdon_samples = int(round(holdon / dt))
+    holdoff_samples = int(round(holdoff / dt))
+    cond = similarity.data >= mean_threshold
+    # loop through True values in cond array and guarantee hold time
+    similarity_cond = similarity.data[cond]
+    cindices = np.nonzero(cond)[0]
+    detections = []
+    i = 0
+    while True:
```
[the remark this answers is not in the document]
Comment (trichter) [reply obspy/signal/cross_correlation.py]: Thanks for pointing out the find_peaks function. It seems suitable. Only the  `holdon` and `holdoff` parameters need to be merged into a single float `distance` parameter, but this is a compromise I am willing to take.

**Comment by trichter** [comment]
> I think this looks really good, and agree that it is great to have this in obspy - I'm looking forward to using it for teaching next year! The tutorial is nice too, although it might be nice to use named arguments in the tutorial so people don't have to flick back and forth between the tutorial and the docs to remind themselves of what each argument means?

Thanks for the review. I agree with the named arguments.

**Comment by trichter** [comment]
Looks good, test failures are unrelated.

**Comment by d-chambers** [comment]
@trichter if @calum-chamberlain  is happy I am happy :+1:

**Comment by trichter** [comment]
I had a second thought about this PR. Here are my conclusions:

* Maybe it is better to merge the three functions again into one. They need to a certain degree the same arguments and it is a bit cumbersome to call all three functions in a row. And maybe it is a bit discouraging for a beginner. The `stream_correlate_template()` function could stay separate, but could be called from the merged function.
* It would be a nice extension to accept not only one template but also a list or dictionary of templates and return the detections with the index or key of the corresponding template.
* I think `templates_max_similarity` can be deprecated, it also does not return the information from which template the best correlation coefficient originates.

**Comment by trichter** [comment]
No objections have been raised, I took this as approval. I thought it was a little change, but ended up refactoring quite a bit. If tests pass and docs show up nicely this PR is ready again.

**Comment by megies** [comment]
> No objections have been raised, I took this as approval. I thought it was a little change, but ended up refactoring quite a bit. If tests pass and docs show up nicely this PR is ready again.

Sorry, didn't have time to give this a proper review yet.. but will do before 1.2.0

**Comment by trichter** [comment]
Arg, doctest is frustrating sometimes.

> Sorry, didn't have time to give this a proper review yet.. but will do before 1.2.0

No problem. When tests pass except for those Windows fails, it is again ready for review and/or merging.

**Comment by trichter** [comment]
PR is ready.

I figured out, that `templates_max_similarity()` is called by `coincidence_trigger()`, but `templates_max_similarity()` seems not to be of much use when called directly. Therefore, I suggest to make the function private and deprecate the original function name.

**Comment by krischer** [comment]
I'm honestly not too deep into the template matching but the three guys on the team who know the most about it gave their green light so I'll press the big button. Thanks a lot for this! its a big addition to ObsPy.

**Comment by krischer** [comment]
> I figured out, that `templates_max_similarity()` is called by `coincidence_trigger()`, but `templates_max_similarity()` seems not to be of much use when called directly. Therefore, I suggest to make the function private and deprecate the original function name.

When was this function introduced? If after 1.1.0 please go ahead with the renaming, otherwise I suggest to wait until after we released 1.2.0.

**Comment by trichter** [comment]
> When was this function introduced? If after 1.1.0 please go ahead with the renaming, otherwise I suggest to wait until after we released 1.2.0.

I think it is pretty old, so let's wait a release.

**Pull request merged**